internal/registry/client_test.go

```diff
@@ -370,3 +372,122 @@ func TestLookupModuleNetworkError(t *testing.T) {
+	}{
+		"shall find the module relying on the OpenTofu registry protocol": {
+			src: "exists-in-registry/identifier/provider",
+		},
+		"shall find the module relying on fallback - OpenTofu alpha registry protocol": {
+			src: "alpha/identifier/provider",
+		},
+		"shall fail to find the module": {
+			src: "not-exist/identifier/provider",
+			// note that the version is fixed in the mock
+			// see: /internal/registry/test/mock_registry.go:testMods
+			wantErrorStr: `module "not-exist/identifier/provider" version "0.2.0" not found`,
+		},
+		"shall fail because of reading response body error": {
+			src:          "foo/bar/baz",
+			wantErrorStr: "error reading response body from registry: foo",
+			httpClient: &http.Client{
+				Transport: mockRoundTripper{
+					v: &http.Response{
+						StatusCode: http.StatusOK,
+						Body:       mockErrorReadCloser{err: errors.New("foo")},
+					},
+				},
+			},
+		},
+		"shall fail to deserialize JSON response": {
+			src:          "foo/bar/baz",
+			wantErrorStr: `module "foo/bar/baz" version "0.2.0" failed to deserialize response body {: unexpected end of JSON input`,
+			httpClient: &http.Client{
+				Transport: mockRoundTripper{
+					v: &http.Response{
+						StatusCode: http.StatusOK,
+						Body:       io.NopCloser(strings.NewReader("{")),
+					},
+				},
+			},
+		},
+		"shall fail because of the registry error": {
+			src:          "foo/bar/baz",
+			wantErrorStr: `error getting download location for "foo/bar/baz": foo resp:bar`,
+			httpClient: &http.Client{
+				Transport: mockRoundTripper{
+					v: &http.Response{
+						// any arbitrary status, but 200, 204 and 404
+						StatusCode: http.StatusFound,
+						Status:     "foo",
+						Body:       io.NopCloser(strings.NewReader("bar")),
+					},
+				},
+			},
+		},
+		"shall fail because location is not found in the response": {
+			src:          "foo/bar/baz",
+			wantErrorStr: `failed to get download URL for "foo/bar/baz": OK resp:{"foo":"git::https://github.com/foo/terraform-baz-bar?ref=v0.2.0"}`,
+			httpClient: &http.Client{
+				Transport: mockRoundTripper{
+					v: &http.Response{
+						StatusCode: http.StatusOK,
+						Status:     "OK",
+						// note that the response emulates a contract change
+						Body: io.NopCloser(strings.NewReader(`{"foo":"git::https://github.com/foo/terraform-baz-bar?ref=v0.2.0"}`)),
+					},
+				},
+			},
+		},
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			client := NewClient(test.Disco(server), tc.httpClient)
+
+			mod, err := regsrc.ParseModuleSource(tc.src)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			_, err = client.ModuleLocation(context.Background(), mod, "0.2.0")
+			if err != nil && tc.wantErrorStr == "" {
+				t.Fatalf("unexpected error: %v", err)
+			}
+			if err != nil && err.Error() != tc.wantErrorStr {
+				t.Fatalf("unexpected error content: want=%s, got=%v", tc.wantErrorStr, err)
+			}
+		})
+	}
+}
+
+type mockRoundTripper struct {
```

Comment:
This looks like a test utility that could be unified somewhere maybe?

Reply:
I feel it's a matter of taste. My preference is to colocate self-made mocks/stubs next to the tests they are dedicated to. Overall, I think that everyone on the @opentofu/core-engineers team shares the feeling refactoring to the codebase would be beneficial after the stable release. I'd scope tests reworking as part of that activity. WDYT?